### DISCUSSION TITLE
#2. Vatz bi-weekly Meeting at 2022-03-02 23:00:00 -06:00:00

### DISCUSSION THREAD

**Discussion by xellos00** [dsrvlabs/vatz#8] [Ideas]
Agenda 

## 1. Vatz Pilot Enhancement.

@dsrvlabs/validator 
Please, leave any comment at least one

**Comment by rootwarp** [comment]
# Discussion items

### Plugin's attaching protocol
According to our first approach, process of adding new plugin is almost done from manager side.
I think this approach takes a lot of complexities for configuration.
Configuration file might be complex and even RDBS might be required.

Moreover, manager should be restarted if we add new plugin or change plugin configuration.
(Sure, this is not necessary if we implement dynamic loading feature.)
It will be better if we operate manager without downtime.

So I propose new plugin protocol to attach plugin features from plugin side itself.
1. Plugin service loaded, it try to connect manager and say `hello`.
2. Manager receive `hello` and transmit response if it available.
3. Plugin and manager starts exchanging features and config parameters
4. Plugin transmits heartbeat signal periodically.

### Plugin interfaces
For now only `Execute` interface provide to invoke plugin feature.
This is simple but too flexible because all information for invoking plugin will be passed by `interface` type.
And that means error handling maybe hard work on plugin side.

For alternatives, `grpc/reflection` package provides reflection interfaces for remote gRPC server.
With that, we can call remote's gRPC API with structured value.

### Manager invoking method
- As I understand, someone should call `manager.Execute` to trigger manager's behavior.
- If I'm right, who call `manager.Execute` and how?
- Is it right to trigger manager with external source?

### Command Line Interface
For management perspective, CLI is one of the powerful tools.

### Fix roles for each components
It might be good start if we discuss about roles for each components to make clear.

Below is just my simple opinions. 
- Manager
  - Collect metrics from plugins.
  - Decide alert situation.
  - Send alert.
- Plugin
  - Provide metrics.
  - Provide special actions(stop, restart, etc.)

**Reply by xellos00** [reply]
@dsrvlabs/validator

I'd simplified previous message, Please put a comment if there's any further ideas regarding previous @rootwarp's message

### 1. new proposal 

1. Plugin service loaded, it try to connect manager and say hello.
2. Manager receive hello and transmit response if it available.
3. Plugin and manager starts exchanging features and config parameters
4. Plugin transmits heartbeat signal periodically

> 1. Can we use init function to do this? or this is passive which nothing has to be call at all?
> 2. I agree, channel #debug? #Alert?
> 3. let's discuss further more on config or feature what's need to be exchanged
> 4. I agree, but is there's specific reason that plugin send a `alive` signal? 

### 2. Manager invoking method

- As I understand, someone should call manager.Execute to trigger manager's behavior.
- If I'm right, who call manager.Execute and how?
- Is it right to trigger manager with external source?

> Seems I missed this part on demo version, but I had put a config with interval automatically trigger execute function previously. Let me put this feature on new development and let's discuss

---

Before Next Monday or previous discussion, I will create a agenda to discuss on our weekly meeting on Vatz

**Comment by hqueue** [comment]
# Discussion item

1. How to maintain multiple plugins in one plugin repository?

Let's discuss an  example, `uptime` plugin and `block height` plugin.
Assumption: above two plugins are maintained in one repo. 
- two plugins should be run independently
- (configuration) configuration for two plugins should be maintained independently.
   - How about configuration of vatz?
- Do we require any common asset for one repo?
   Nope. Each plugins should be maintained independently.
- many part of source code may be duplicated (can we make it simpler?)
   ```
   # Suggested structure
   vats-plugin-xxx/Makefile
   vats-plugin-xxx/cmd/xxx-uptime.go
   vats-plugin-xxx/cmd/xxx-blockheight.go
   vats-plugin-xxx/uptime/*
   vats-plugin-xxx/blockheight/*
   vats-plugin-xxx/plugin/* (protobuf)
   ...
   ```
- (build/compile)two plugins should be compiled independently

2. Configuration file format? `yaml`, `toml`, `json` and etc.
   - 논의: `yaml` (합의)

**Comment by hqueue** [comment]
## 형기님 리뷰
- 현재 준교님 코드 기준으로는 새로운 plugin 개발은 하루면 충분할 듯합니다
- 빈칸채우기가 되면 더욱 좋을 듯

## 희진님 리뷰
Q. Plugin Execute는 하나의 status/metric만 확인해야하는가?
   A: 하나의 plugin은 하나의 status/metirc만 확인하는 것이 좋아보인다.

Q. Plugin Execute가 호출되는 주기(period)는?
   A: Vatz가 호출 주기는 설정파일로 결정하는 방식으로 계획되어 있다

Q. Plugin의 return한 결과를 Vatz는 어떻게 처리해야하는가?
   A. Vatz가 처리하면 될 듯하다 (설정 어떻게 관리할까?)

## 진홍
- 현재 준교님 코드 기준으로는 새로운 plugin 개발은 하루면 충분할 듯합니다
- 빈칸채우기가 되면 더욱 좋을 듯

**Comment by hqueue** [comment]
# 2022-03-02 논의된 내용

1. plugin은 하나의 역할만 수행
2. Configuration file은 모두 vatz에서 관리한다. `yaml` 이용
3. vatz가 시작 시 노드 운영자가 설정 파일을 지정(url, github, local file 등) 할 수 있도록 한다
